language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -59,17 +70,21 @@ added to environments and normalization instantiates all variables presented in
 -- | Normalize a 'Type' in the 'NormalizeTypeM' monad.
 normalizeTypeM
     :: HasUnique (tyname ()) TypeUnique
-    => Type tyname () -> NormalizeTypeM tyname (NormalizedType tyname ())
-normalizeTypeM (TyForall ann name kind body) = TyForall ann name kind <<$>> normalizeTypeM body
-normalizeTypeM (TyFix ann name pat)          = TyFix ann name <<$>> normalizeTypeM pat
+    => Type tyname () -> NormalizeTypeM tyname a (NormalizedType tyname ())
+normalizeTypeM (TyForall ann name kind body) =
+    TyForall ann name kind <<$>> normalizeTypeM body
+normalizeTypeM (TyFix ann name pat)          =
+    TyFix ann name <<$>> normalizeTypeM pat
 normalizeTypeM (TyFun ann dom cod)           =
     TyFun ann <<$>> normalizeTypeM dom <<*>> normalizeTypeM cod
-normalizeTypeM (TyLam ann name kind body)    = TyLam ann name kind <<$>> normalizeTypeM body
+normalizeTypeM (TyLam ann name kind body)    =
+    TyLam ann name kind <<$>> normalizeTypeM body
 normalizeTypeM (TyApp ann fun arg)           = do
     vFun <- normalizeTypeM fun
     vArg <- normalizeTypeM arg
     case getNormalizedType vFun of
-        TyLam _ nArg _ body -> substituteNormalizeTypeM vArg nArg body
+        -- TODO: we should check nArg more thoroughly for non-normal types
+        TyLam _ nArg _ body -> normalizeTypeStep *> substituteNormalizeTypeM vArg nArg body
```

Comment:
Wouldn't it be simpler to put `normalizeTypeStep` inside `substitute`, since that's what we're tracking?

Reply:
We don't actually want this to do anything with types that are already normalized. `substituteNormalizeType` is called in `Language.PlutusCore.TypeSynthesis`.

Reply:
Hm, then it seems to me that our specification of what we're tracking is underspecified. Are we tracking all substitutions (some of which happen e.g. when dealing with `forall`s), or are we only tracking substitutions that happen as a result of type lambda reductions? I think we do implicitly mean the latter, so this is right, but we should document it carefully.